code.js

```diff
@@ -1,3 +1,100 @@
 function tsp_ls(distance_matrix) {
-    return -1;
+    var path = Array.from(Array(distance_matrix.length).keys())
+    path = path.reduce(
+        (acc, _, i) => {
+            random = Math.floor(Math.random() * (acc.length - i)) + i;
+            [acc[i], acc[random]] = [acc[random], acc[i]]; 
+            return acc;
+        },
+        [...path]
+    );
+    var current_distance = Infinity
+    var best_distance = costOfPath(distance_matrix,path)
+    do{ improvementMade = false //have to run through the first iteration
+        for(let i = 1; i < path.length-2; i++){ 
+            for(let j = i+1; j < path.length-1; j++){ //based on the observations made in https://en.wikipedia.org/wiki/2-opt
+                path = TwoOptSwap(path, i, j)
+                //console.log(path + " " + i + " " + j)
+                current_distance = costOfPath(distance_matrix,path)
+                if(current_distance < best_distance){ //stopping criteria, no improvement, no need to keep going
+                    best_distance = current_distance
+                    improvementMade = true;
+                    break;
+                }
+            }
+            if(improvementMade) break; //really wish I could just use goto
```

Comment:
Don't understand your logic here -- you're stopping if you're finding an improving route?